Improve model qc

xylempalace opened the pull request:
Improved QC by adding 2 new plotting functions for comparing synthetic and observed data.

xylempalace committed:
Feat: Added plot_marginal_numerical_numerical function which plots 2 numerical data sets in a scatter plot

xylempalace committed:
feat: Added plot_marginal_numerical_categorical function shows numerical and categorical data as box plots from 2 different data sets

xylempalace commented:
a

Schaechtle commented on jupyter/marginal-1d-plot.ipynb:
```diff
@@ -14,7 +14,7 @@
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": null,
```
run `git checkout [path-to-this-file]`. It didn't actually change other than for metadata.

Schaechtle replied:
@xylempalace , can we fix this ^

Schaechtle commented on jupyter/marginal-num-cat-plot.ipynb:
This example crashes for me.

xylempalace replied:
Fixed it and added docstrings

Schaechtle commented on src/lpm_plot/plot_marginal.py:
```diff
@@ -153,3 +153,55 @@ def plot_marginal_2d(combined_df, x, y, hm_order=None, cmap="oranges"):
     return combined_heatmap
+
+
+def plot_marginal_numerical_numerical(observed_df, synthetic_df, x, y):
```
please add docstrings

Schaechtle reviewed (changes requested):
The code looks reasonable to me but I wanted to run the numerical categorical notebook example and it crashes. Please fix.

Also, for any numerical axis, please provide optional arguments for axis-limits (that wasn't in my original spec but is very important).

xylempalace committed:
feat!: add new parameter detailed_df to plot_heatmap which displays a new graph that displays more info about the relationship between a selected heatmap cell

xylempalace committed:
fix: Fixed detail graph labels for plot_heatmap

xylempalace committed:
fix!: Fixed issues with example for plot_marginal_numerical_numerical and added new params x_domain and y_domain. Also added documentation

xylempalace commented:
Changed plot_heatmap function to now be able to display detailed data when clicking a cell on the heat map if the function is provided with more data

Schaechtle commented on src/lpm_plot/plot_marginal.py:
```diff
@@ -153,3 +153,131 @@ def plot_marginal_2d(combined_df, x, y, hm_order=None, cmap="oranges"):
+    x: str,
+    y: str,
+    x_domain: tuple[float, float] = [0.0, 10.0],
+    y_domain: tuple[float, float] = [0.0, 10.0],
```
That's not a good default for the x and y axis limits. Let's default to `None` and add a condition that if `x_domain`, or `y_domain` is `None` we take the min and max values from `observed_df` to set the limits/domain.

npovejsil replied:
I believe this is addressed

Schaechtle commented on src/lpm_plot/plot_marginal.py:
```diff
@@ -153,3 +153,131 @@ def plot_marginal_2d(combined_df, x, y, hm_order=None, cmap="oranges"):
+    synthetic_df: pl.DataFrame,
+    x: str,
+    y: str,
+    size: float = 30.0,
```
let's add an optional y_domain argument, here, too, like above.

npovejsil replied:
addressed

Schaechtle commented on src/lpm_plot/plot_marginal.py:
```diff
@@ -153,3 +153,131 @@ def plot_marginal_2d(combined_df, x, y, hm_order=None, cmap="oranges"):
+    Args:
+        observed_df : pl.DataFrame
+            A Polars DataFrame containing the observed data and it must contain the columns named
+            after each category and are filled with the data points for that category
```
I don't think I understand the English here / what you mean

npovejsil replied:
addressed

Schaechtle reviewed (changes requested):
@xylempalace, this is good.

Here is some feedback.

### 1. Categorical input to boxplot and categorical-categorical plots in `plot_heatmap`.
The boxplot's visuals are correct, but the input dataframes are not (that's `plot_marginal_numerical_categorical`. In your example notebook, I am seeing that categoricals 
```
+----------+----------+
|   foo    |   bar    |
+----------+----------+
|   0.125  |  apple   |
|   2.500  |  banana  |
|  10.000  |  cherry  |
|   ...    |   ...    |
|   4.200  |  apple   |
+----------+----------+
```
Same for `plot_heatmap`. There is no "heats score", we compute the frequency of the events observed in the dataframe on the fly.  Assume the following dataframe:
```
+---------+------+
|  cat1   | cat2 |
+---------+------+
|  foo    | yes  |
|  bar    | no   |
|  quagga | yes  |
|  ...    | ...  |
|  foo    | no   |
+---------+------+
```
An event, in this case, is every possible combination of the values in the two categorical columns (`foo, yes`, `foo, no`, ...). And of course, we want to update the colorbar label to "frequency". Be careful to ensure that missing events have a frequency of zero and are not `None`.

### 2. Start-up of interactive MI looks messy

Somehow, I am getting a weirdly overlapping default when I start `plot_heatmap`
before clicking anything. See this screenshot <img width="1139" height="968" alt="image" src="https://github.com/user-attachments/assets/8bf8f5df-e1b1-42cc-9daf-e703ed4ce3ae" />.


### 3. Color scales

Let's leave the green heatmap for the main heatmap from `plot_heatmap` but change the other colors to black greyscale for the plots that appear on click. Let's use black for the scatter plot and the boxplot and Greyscales for the heatmap. Otherwise, it's visually too much.

### 4. Boxplot axis labels

The labels seem swapped -- see screenshot below.
<img width="1120" height="968" alt="image" src="https://github.com/user-attachments/assets/000983e1-070c-4d1c-975e-aafaf16dd651" />

### 5. Superfluous colorbar

The colorbar in the screenshot below should not be there. 
<img width="1120" height="968" alt="image" src="https://github.com/user-attachments/assets/f1dcbe25-6225-4b67-8268-5f691863db20" />

### 6. Other

I've also left a bunch of comments to be addressed in-line.

### 7. Git rebase

`git rebase` the two fixes into your original `feat` commits. If you haven't used `git rebase` before, please ask your dad for help. `git` is a complex system, and `git rebase` is a dangerous tool.

xylempalace commented:
@Schaechtle 
> ### 4. Boxplot axis labels
> The labels seem swapped -- see screenshot below.

This was done on purpose to avoid axis overlap issues for numerical vs numerical and categorical vs categorical comparisons. I'll leave a note in the code.

> ### 5. Superfluous colorbar
> The colorbar in the screenshot below should not be there.

To my knowledge it is impossible to get rid of this.

Schaechtle commented:
> @Schaechtle
> 
> > ### 4. Boxplot axis labels
> > The labels seem swapped -- see screenshot below.
> 
> This was done on purpose to avoid axis overlap issues for numerical vs numerical and categorical vs categorical comparisons. I'll leave a note in the code.
> 
well, but they are wrong, right? the x axis is labelled as the y-axis should? We can't leave that be. 

> > ### 5. Superfluous colorbar
> > The colorbar in the screenshot below should not be there.
> 
> To my knowledge it is impossible to get rid of this.

That sucks. Is there an open issue against vega-lite? If not, we should consider opening one.

xylempalace commented:
> > @Schaechtle
> > > ### 4. Boxplot axis labels
> > > The labels seem swapped -- see screenshot below.
> > 
> > 
> > This was done on purpose to avoid axis overlap issues for numerical vs numerical and categorical vs categorical comparisons. I'll leave a note in the code.
> 
> well, but they are wrong, right? the x axis is labelled as the y-axis should? We can't leave that be.
I see what happened I actually misread what you originally brought up. I can totally fix that.
> > > ### 5. Superfluous colorbar
> > > The colorbar in the screenshot below should not be there.
> > 
> > 
> > To my knowledge it is impossible to get rid of this.
> 
> That sucks. Is there an open issue against vega-lite? If not, we should consider opening one.

The issue is that in order to make a graph that changes its actually the scatter plot, box plot, and heat map all on top of each other and when a user clicks it hides certain graphs and shows the relevant graph and apparently the legend for the heatmap doesn't get filtered out. I haven't found an open issue against it either.

xylempalace committed:
feat!: improve heatmap and marginal plot customization

- Refactor heatmap color parameters from tuple to separate cmap_main, cmap_detail, and detail_color options
- Enhance marginal numerical plot domain handling to auto-calculate from data when None provided
- Update selection handling in heatmap for better user interaction
- Update test data and notebooks to reflect API changes

xylempalace committed:
feat!: add interactive heatmap demo and improve zero-value visualization

- Add new heatmap-interaction-demo.ipynb with sample datasets
- Improve heatmap color handling to display zero values as white
- Add mutual information and training data parquet files for testing
- Update version and enhance existing notebook examples

xylempalace committed:
feat: Add demo data files, update heatmap examples, and add comments
- Add demo-heatmap-df.parquet and demo-detail-df.parquet for example datasets

npovejsil commented:
@xylempalace It's looking good! Some small fixes below. 

Feedback:

Issues with _plot_heatmap.py, heatmap-interation-demo.ipynb, heatmap-plot.ipynb_
- [ ] Implement similar handling to the "No Data: self comparison" case for null/0 combinations of data. Try to get rid of the frequency bar in those cases too. Screenshot from _heatmap-interaction-demo.ipynb_
<img width="573" height="369" alt="Screenshot 2025-08-29 at 11 22 12 AM" src="https://github.com/user-attachments/assets/7a954916-1889-43e1-9b14-c53bf8cd5def" />

- [ ] Remove frequency bar on the right for scatterplots (numeric-numeric comparison, irrelevant to show frequency). Screenshot from _heatmap-plot.ipynb_
<img width="647" height="577" alt="Screenshot 2025-08-29 at 11 25 14 AM" src="https://github.com/user-attachments/assets/97c18ba3-cf40-48ee-b968-31ddd178fe09" />

- [ ]  If possible, move the number ticks to the bottom of the plot for readability for the num/cat combinations (also a bit confused about the 0 on the right-hand side ): 
<img width="674" height="547" alt="Screenshot 2025-08-29 at 11 36 54 AM" src="https://github.com/user-attachments/assets/50594c1f-fbd6-4ad0-a280-bc124326277d" />


Issue with _plot_marginal.py_ and _marginal-num-num-plot.ipynb_: 

- [ ] Add a legend for synthetic vs. real (or whatever the color breakdown is) for the scatterplot 

<img width="572" height="423" alt="Screenshot 2025-08-29 at 11 32 53 AM" src="https://github.com/user-attachments/assets/06465ba6-009f-45a5-81a2-8e94330ba01c" />

xylempalace committed:
fix: Added new empty graph to plot_heatmap and added legend to plot_marginal_numerical_numerical

xylempalace commented:
@npovejsil Addressed the feedback. The only one I was actually able to address was the last one. The first issue would cause performance issues if I were to resolve it. The frequency bar is unable to be hidden. The reason the number ticks are swapped is because when the bar graph is filtered out it shows up as the 0 ticks which I can't get rid of so I put those number ticks on the opposite side to avoid overlapping with other labels.

npovejsil commented:
@xylempalace Cool! If there are problems that still need to be addressed but there's some bug with Vega-lite, see if you can submit an issue describing what the bug is here: [](https://github.com/vega/vega-lite). Reach out if you get stuck during that process! There might be on-going conversations/issues already pertaining to the problems you're having that you can check out too and see if there's any advice people have. 

Can you expand on the performance issues? Does that mean it's so slow to plot? 

The ticks being on that side, in my opinion, is fine but you can check with Ulli.
